lib/logger.js

```diff
@@ -1,456 +1,634 @@
-  , env: this.source.env
-  , meta: this._meta
+  flush() {
+    // Roll the current buffer into readyToSend and begin a new buffer
+    clearTimeout(this[kFlusher])
+    this[kFlusher] = null
+
+    const bufferLength = this[kBuffer].length
+    if (bufferLength) {
+      this[kReadyToSend].push(this[kBuffer])
+      this[kBuffer] = []
+      this[kLineLengthTotal] = 0
+      this[kTotalLinesReady] += bufferLength
+    }
+
+    if (this[kReadyToSend].length) {
+      this.send()
+      return
+    }
+
+    // setImmediate allows us to flush THEN await the event in code
+    setImmediate(this.emit.bind(this), 'cleared', {
+      message: ALL_CLEAR_EMPTY
+    })
   }
 
-  if (opts) {
-    if (typeof opts === 'function') {
-      callback = opts
-    } else if (typeof opts === 'string') {
-      if (opts.length > configs.MAX_INPUT_LENGTH) {
-        debug('Level had more than ' + configs.MAX_INPUT_LENGTH + ' chars, was truncated')
-        opts = opts.substring(0, configs.MAX_INPUT_LENGTH)
-      }
-      message.level = opts
-    } else {
-      if (typeof opts !== 'object') {
-        const err = new TypeError('options parameter must be a level (string), or object')
-        err.meta = {opts}
-        this._err = err
-        debug(err.message)
-      }
+  log(statement, opts) {
+    opts = opts || {}
 
-      message.level = opts.level || message.level
-      message.app = opts.app || message.app
-      message.env = opts.env || message.env
+    if (statement === null
+      || statement === undefined
+      || typeof statement === 'string' && !statement.length) {
 
-      if (opts.timestamp && isValidTimestamp(opts.timestamp)) {
-        message.timestamp = opts.timestamp
-      }
+      this.emit('warn', {
+        message: WARN_LOG_IGNORED
+      , statement
+      })
+      return
+    }
 
-      if (opts.context && !opts.meta) {
-        opts.meta = opts.context
-      }
+    const message = {
+      timestamp: Date.now()
+    , line: undefined
+    , level: this.level
+    , app: this.app
+    , env: this.env
+    , meta: undefined
+    }
 
-      if (typeof opts.meta === 'object') {
-        opts.meta = {...message.meta, ...opts.meta}
-        if (opts.index_meta || (opts.index_meta === undefined && this._index_meta)) {
-          message.meta = opts.meta
-        } else {
-          message.meta = stringify(opts.meta)
+    if (typeof opts === 'string') {
+      // Then it should be a log level string
+      const level = opts.toUpperCase()
+      if (!constants.LOG_LEVELS.includes(level)) {
+        const err = new TypeError(
+          'If \'options\' is a string, then it must be a valid log level'
+        )
+        err.meta = {
+          got: opts
+        , expected: constants.LOG_LEVELS
         }
+        this.emit('error', err)
+        return
       }
-
-      if (opts.logSourceCRN) {
-        message.logSourceCRN = opts.logSourceCRN
+      message.level = level
+      opts = {}
+    }
+    if (typeof opts !== 'object') {
```

Comment:
might consider pulling in `typeOf` for this check (and others) since this is a public interface

Reply:
oh, I certainly did, and I contemplated that :]. The only thing I think it gives us is discerning array from object.  Since we both thought of it, makes sense to add it in.